### PR TITLE
feat: put transcript

### PR DESCRIPTION
**Pull request by micael-jerry**

### CODE REVIEW

File: src/main/java/school/hei/haapi/endpoint/rest/controller/TranscriptController.java [hunk 1]
```diff
@@ -1,12 +1,10 @@
 package school.hei.haapi.endpoint.rest.controller;
 
 import lombok.AllArgsConstructor;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.RequestParam;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (fenohasinalala) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/TranscriptController.java]: Tsy mety ilay .* fa mila specifierna daholo ilay dependance ilaina

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/TranscriptMapper.java [hunk 2]
```diff
@@ -16,4 +20,19 @@ public Transcript toRest(school.hei.haapi.model.Transcript transcript) {
...
+        if (student == null) {
+            throw new NotFoundException("Student.id=" + studentId + " is not found");
+        }
+        return school.hei.haapi.model.Transcript.builder()
+                .id(transcript.getId())
+                .student(student)
+                .semester(school.hei.haapi.model.Transcript.Semester.valueOf(transcript.getSemester().toString()))
+                .academicYear(transcript.getAcademicYear())
+                .isDefinitive(transcript.getIsDefinitive())
+                .creationDatetime(transcript.getCreationDatetime())
```
Comment (fenohasinalala) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/TranscriptMapper.java:35]: ilay creationDateTime mety asina kely hoe raha id tsy mbola misy any @db ilay transcript (id== null) dia egale Instant.now()
dia raha id == null dia isDefinitve false (sa efa misy default false ilay shemat sql?)

File: src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java [hunk 3]
```diff
@@ -113,6 +113,7 @@ protected void configure(HttpSecurity http) throws Exception {
         .requestMatchers(new SelfMatcher(GET, "/students/*/transcripts/*")).hasAnyRole(STUDENT.getRole())
         .antMatchers(GET, "/students/*/transcripts").hasAnyRole(TEACHER.getRole(), MANAGER.getRole())
         .antMatchers(GET, "/students/*/transcripts/*").hasAnyRole(TEACHER.getRole(), MANAGER.getRole())
+            .antMatchers(PUT, "/students/*/transcripts").hasAnyRole(TEACHER.getRole(), MANAGER.getRole())
```
Comment (fenohasinalala) [inline src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java]: Mila alignena @ hafa ilay izy

File: src/main/java/school/hei/haapi/endpoint/rest/validator/CreateTranscriptValidator.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+import java.util.List;
+import java.util.function.Consumer;
+
+@Component
+public class CreateTranscriptValidator implements Consumer<Transcript> {
+
+    @Override
+    public void accept(Transcript transcript) {
+        if (transcript.getId() == null) {
```
Comment (fenohasinalala) [inline src/main/java/school/hei/haapi/endpoint/rest/validator/CreateTranscriptValidator.java:15]: ilay id possible null satria tsy mitovy @ ilay Claim hoe foron'ny Client a la creation.
Zany hoe tsy atao io validation 1 io.
Fa ilay semester koa no tokony mandatory, (ilay academic_year possible asiana koa)

File: src/main/java/school/hei/haapi/repository/TranscriptRepository.java [hunk 5]
```diff
@@ -14,4 +14,6 @@ public interface TranscriptRepository extends JpaRepository<Transcript, String>
 
     Optional<Transcript> getByIdAndStudentId(String id, String studentId);
 
+    Transcript findByIdAndStudentId(String id, String studentId);
```
Comment (fenohasinalala) [inline src/main/java/school/hei/haapi/repository/TranscriptRepository.java]: Efa misy ilay query mitovy amininy lesy io ambony io. dia ilay service mampiasa an'ilay efa ao no ampiasaina.
tsy mila manao query hafa zany

File: src/main/java/school/hei/haapi/service/TranscriptService.java [hunk 6]
```diff
@@ -28,4 +33,26 @@ public List<Transcript> getAllByStudentId(String studentId, PageFromOne page, Bo
...
+
+    public List<Transcript> updateStudentTranscript(String studentId, List<Transcript> transcripts) {
+        User user = userRepository.getById(studentId);
+        return transcripts.stream()
+                .map(transcript -> this.updateTranscript(user, transcript))
+                .collect(Collectors.toList());
+    }
+
+    private Transcript updateTranscript(User user, Transcript update) {
+        Transcript studentTranscript = transcriptRepository.findByIdAndStudentId(update.getId(), user.getId());
```
Comment (fenohasinalala) [inline src/main/java/school/hei/haapi/service/TranscriptService.java]: io ilay ampiasaina lesy 
public Transcript getByIdAndStudent(String id,String studentId)

File: src/main/java/school/hei/haapi/service/TranscriptService.java [hunk 7]
```diff
@@ -28,4 +33,26 @@ public List<Transcript> getAllByStudentId(String studentId, PageFromOne page, Bo
...
+                    .student(user)
+                    .creationDatetime(update.getCreationDatetime())
+                    .build();
+        }
+        studentTranscript.setSemester(update.getSemester());
+        studentTranscript.setAcademicYear(update.getAcademicYear());
+        studentTranscript.setIsDefinitive(update.getIsDefinitive());
+        return transcriptRepository.save(studentTranscript);
+    }
 }
```
Comment (fenohasinalala) [inline src/main/java/school/hei/haapi/service/TranscriptService.java:62]: Dia mila second validation koa avant miditra ao @ bd mila mijery exemple ao @ ilay package /model/validator

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/TranscriptMapper.java (hunk 2 above)
Comment (micael-jerry) [reply src/main/java/school/hei/haapi/endpoint/rest/mapper/TranscriptMapper.java:35]: efa mis default now() ny creationDatetime
de ef mis default ko ny isDefinitive

**Pull request merged**